src/capture/capture_video_ffmpeg.cpp

```diff
@@ -0,0 +1,681 @@
+
+// Disable deprecated warnigns coming from ffmpeg libraries.
+// I'm trying to support as many versions as possible.
+// Some of the CI machines don't support the new functions.
+// And some give deprecated warnings on the new functions.
+#ifdef _MSC_VER
+#pragma warning(disable:4996)
+#else
+// This also works on Clang
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+#endif
+
+// These get added to an array in muxer.format_context->streams
+// In the order they get initalized with avformat_new_stream()
+constexpr int MuxerVideoStreamIndex = 0;
+constexpr int MuxerAudioStreamIndex = 1;
+
+FFmpegEncoder::FFmpegEncoder()
+{
+	audio_encoder.thread = std::thread(&FFmpegEncoder::encode_audio, this);
+	video_encoder.thread = std::thread(&FFmpegEncoder::encode_video, this);
+	muxer.thread         = std::thread(&FFmpegEncoder::mux, this);
+}
+
+FFmpegEncoder::~FFmpegEncoder()
+{
+	mutex.lock();
+	worker_threads_should_quit = true;
+	mutex.unlock();
+	waiter.notify_all();
+	stop_queues();
+	if (audio_encoder.thread.joinable()) {
+		audio_encoder.thread.join();
+	}
+	if (video_encoder.thread.joinable()) {
+		video_encoder.thread.join();
+	}
+	if (muxer.thread.joinable()) {
+		muxer.thread.join();
+	}
+}
+
+static AVPixelFormat map_pixel_format(PixelFormat format)
+{
+	switch (format) {
+	case PixelFormat::Indexed8:
+		// TODO
+		break;
+	case PixelFormat::BGR555:
+#ifdef WORDS_BIGENDIAN
+		return AV_PIX_FMT_BGR555BE;
+#else
+		return AV_PIX_FMT_BGR555LE;
+#endif
+	case PixelFormat::BGR565:
+#ifdef WORDS_BIGENDIAN
+		return AV_PIX_FMT_BGR555BE;
+#else
+		return AV_PIX_FMT_BGR555LE;
+#endif
+	case PixelFormat::BGR888: return AV_PIX_FMT_BGR24;
+	case PixelFormat::BGRX8888: return AV_PIX_FMT_BGR0;
+	}
+	return AV_PIX_FMT_NONE;
+}
+
+void FFmpegEncoder::capture_video_add_frame(const RenderedImage& image,
+                                            const float frames_per_second)
+{
+	const int rounded_fps = static_cast<int>(frames_per_second + 0.5f);
+	if (!video_encoder.is_initalised) {
+		if (!video_encoder.init(image.width,
+		                        image.height,
+		                        image.pixel_format,
+		                        rounded_fps)) {
+			LOG_ERR("FFmpeg: Video encoder failed to init");
+			return;
+		}
+		bool init_muxer = false;
+		if (!muxer.is_initalised && audio_encoder.is_initalised) {
+			init_muxer = muxer.init(video_encoder, audio_encoder);
+			if (!init_muxer) {
+				LOG_ERR("FFmpeg: Failed to init muxer");
+				video_encoder.free();
+				return;
+			}
+		}
+		mutex.lock();
+		video_encoder.is_initalised = true;
+		muxer.is_initalised         = muxer.is_initalised || init_muxer;
+		mutex.unlock();
+		waiter.notify_all();
+	} else if (video_encoder.width != image.width ||
+	           video_encoder.height != image.height ||
+	           video_encoder.pixel_format != image.pixel_format ||
+	           video_encoder.frames_per_second != rounded_fps) {
+		capture_video_finalise();
+		if (!video_encoder.init(image.width,
```

Comment:
Some blank lines would help to break up this "block of text" look and make visual parsing easier 😅